packages/tools/viewer-alpha/src/viewerElement.ts

```diff
@@ -0,0 +1,106 @@
+ */
+export class HTML3DElement extends HTMLElement {
+    // eslint-disable-next-line jsdoc/require-jsdoc, @typescript-eslint/naming-convention
+    public static readonly observedAttributes = Object.freeze(["src", "env"] as const);
+
+    private readonly _viewer: Viewer;
+
+    /**
+     * Creates an instance of HTML3DElement.
+     */
+    public constructor() {
+        super();
+
+        const shadowRoot = this.attachShadow({ mode: "open" });
+        shadowRoot.innerHTML = `
+          <style>
+            :host {
+              display: block;
+              width: 100%;
+              height: 100%;
+            }
+
+            #container {
+              display: block;
+              width: 100%;
+              height: 100%;
+            }
+
+            #renderCanvas {
+              width: 100%;
+              height: 100%;
+              display: block;
+              font-size: 0;
+            }
+          </style>
+          <div id="container">
+            <canvas id="renderCanvas" touch-action="none"></canvas>
+          </div>
+        `;
+
+        const canvas = shadowRoot.querySelector("#renderCanvas") as HTMLCanvasElement;
+        this._viewer = createViewerForCanvas(canvas);
+    }
+
+    /**
+     * Gets the model URL.
+     */
+    public get src() {
+        return this.getAttribute("src");
+    }
+
+    /**
+     * Sets the model URL.
+     */
+    public set src(value: Nullable<string>) {
+        if (value === null) {
+            this.removeAttribute("src");
+        } else {
+            this.setAttribute("src", value);
+        }
+    }
+
+    /**
+     * Called each time the element is added to the document.
+     * @remarks
+     * See https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_element_lifecycle_callbacks
+     */
+    public connectedCallback() {}
+
+    /**
+     * Called when attributes are changed, added, removed, or replaced.
+     * @remarks
+     * See https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_element_lifecycle_callbacks
+     * @param name The name of the attribute that changed.
+     * @param oldValue The old value of the attribute.
+     * @param newValue The new value of the attribute.
+     */
+    public attributeChangedCallback(name: (typeof HTML3DElement.observedAttributes)[number], oldValue: string, newValue: string) {
+        switch (name) {
+            case "src":
+                this._viewer.loadModelAsync(newValue).catch(Logger.Error);
+                break;
+            case "env":
+                this._viewer.loadEnvironmentAsync(newValue).catch(Logger.Error);
+                break;
+        }
+    }
+}
+
+/**
+ * Registers all Viewer related custom elements.
+ * @remarks
+ * This function must be called once before using any Viewer related custom elements.
+ */
+export function registerCustomElements() {

```

Comment:
It seems common for custom elements to call `customElements.define` as a side effect, but I opted for this to be an explicit function call to avoid the pain side effects cause for bundlers. If you check the test app's `index.html` in the public directory, you can see that it just requires the consumer to call this function when using the `<babylon-viewer>` custom element. Anyone have any thoughts on this?

Reply:
Does feel like a case for side effects, but importing the function and executing it makes sense as well. I am very neutral about this one TBH. It does mean, however, that every user will have to call this function, right? Could it be done in the viewer's constructor, for example? Or will it fail if not executed before?

Reply:
Can't be done in the constructor as the constructor won't be invoked until the custom element is defined (e.g. the browser knows to map `babylon-viewer` DOM elements to the `HTML3DElement` class.

We need to investigate this more to understand if there is any way to have the registration automatically happen without depending on side effects. Seems like other popular packages may achieve this somehow, but it's not yet clear how.

Reply:
We'll continue this investigation but not block this PR on it.